File: interceptors/lunar-ts-interceptor/src/trafficFilter.ts
import { logger } from './logger'
import { type OutgoingHttpHeaders } from "http"
import { popHeaderValue } from "./helper"

const LIST_DELIMITER = ","
const ALLOW_LIST = "AllowList"
const BLOCK_LIST = "BlockList"
const ALLOWED_HEADER_KEY = "x-lunar-allow"
const IP_PATTERN = /^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$/;

export class TrafficFilter {
    private _managed: boolean = false
    private _stateOk: boolean = false
    private readonly _isExternalIPCache: Map<string, boolean> = new Map<string, boolean>()
    private _allowList: string[] | undefined
    private _blockList: string[] | undefined

    public constructor() {
        this._allowList = this.parseList(process.env["LUNAR_ALLOW_LIST"])
        this._blockList = this.parseList(process.env["LUNAR_BLOCK_LIST"])
    }

    private checkIfHostOrIPIsAllowed(hostOrIp: string): boolean {
        const isAllowed = this.checkAllowed(hostOrIp)

        if (isAllowed !== undefined) return isAllowed

        return this.checkBlocked(hostOrIp) && this.isExternalIP(hostOrIp)
    }

    private checkAllowed(hostOrIp: string): boolean | undefined {
        if (this._allowList === undefined) return undefined

        return this._allowList.includes(hostOrIp)
    }

    private checkBlocked(hostOrIp: string): boolean {
        if (this._blockList === undefined) return true

        else if (this._blockList.includes(hostOrIp)) return false

        return true
    }

    private parseList(valueToParse?: string): string[] | undefined {
        if (valueToParse === undefined) return undefined

        return valueToParse.split(LIST_DELIMITER)
    }

    private isAccessListValid(): boolean {
        if (!this.validateAllow()) return false

        if (!this.validateBlock()) {
            logger.warn("Interceptor will be disable to avoid passing wrong traffic through the Proxy.")
            return false
        }

        return true
    }

    private validateAllow(): boolean {
        if (this._allowList === undefined) {
            return true
        }

        const valuesToRemove: string[] = []
        for (const hostOrIP of Object.values(this._allowList)) {
            if (!(this.validateHost(hostOrIP) || this.validateIP(hostOrIP))) valuesToRemove.push(hostOrIP)
        }

        logger.warn(`Unsupported value ['${valuesToRemove.toString()}'] will be removed from the allowed list.`)

        this._allowList = this._allowList.filter(item => !valuesToRemove.includes(item))
        return true
    }

    private isExternalIP(ip: string): boolean {
        if (this._isExternalIPCache.has(ip)) return this._isExternalIPCache.get(ip) as boolean
        if (!this.validateIP(ip)) return true // If it's not an IP, we currently assume it's external
        const octets = ip.split('.').map((octet: string) => parseInt(octet));
        let isExternal = false
        if (octets[0] === 10) {
            // 10.x.x.x is a private IP
        } else if (octets[0] === 172 && octets[1] !== undefined && octets[1] >= 16 && octets[1] <= 31) {
            // 172.16.x.x - 172.31.x.x is a private IP
        } else if (octets[0] === 192 && octets[1] === 168) {
            // 192.168.x.x is a private IP
        } else if (octets[0] === 169 && octets[1] === 254) {
            // 169.254.x.x is a private IP
        } else if (octets[0] === 127) {
            // 127.x.x.x is a loopback IP
        } else isExternal = true; // Not an internal IP
        
        this._isExternalIPCache.set(ip, isExternal)
        return isExternal
    }

    private validateBlock(): boolean {
        if (this._blockList === undefined) return true

        if (this._allowList !== undefined) {
            logger.warn(`TrafficFilter::Found ${ALLOW_LIST} skipping the ${BLOCK_LIST}`)
            this._blockList = []
            return true
        }

        let blockListValidatePass: boolean = true
        for (const hostOrIP of Object.values(this._blockList)) {
            if (!(this.validateHost(hostOrIP) || this.validateIP(hostOrIP))) {
                logger.warn(`Error while parsing '${hostOrIP}' from the block list`)
                blockListValidatePass = false
            }
        }

        return blockListValidatePass
    }

    private validateHost(host: string): boolean {
        if (this.validateIP(host)) return false
        return true
    }

    private validateIP(ip: string): boolean {
        return IP_PATTERN.test(ip)
    }

    public setManaged(isManaged: boolean): void {
        this._managed = isManaged
        this._stateOk = this.isAccessListValid()
    }

    public isAllowed(hostOrIp: string, headers?: OutgoingHttpHeaders): boolean {
        const hostOrIPWithoutPort = hostOrIp.split(":")[0]
        logger.debug(`TrafficFilter::Checking if called to destination: "${hostOrIPWithoutPort}" is allowed.`)
        if (!this._stateOk) return false
        let isHeaderFilteredAllowed = true

        const isAllowedByHeader = popHeaderValue(ALLOWED_HEADER_KEY, headers);
        if (logger.isDebugEnabled() && isAllowedByHeader !== undefined) {
            logger.debug(`TrafficFilter::Filtering url: "${hostOrIPWithoutPort}" by header ${ALLOWED_HEADER_KEY}=${isAllowedByHeader as string}`);
        }
        
        if (isAllowedByHeader !== undefined) isHeaderFilteredAllowed = isAllowedByHeader === "true";
        
        return isHeaderFilteredAllowed && this.checkIfHostOrIPIsAllowed(hostOrIPWithoutPort ?? '');
    }

    public isManaged(): boolean {
        return this._managed;
    }
}

File: interceptors/lunar-py-interceptor/lunar_interceptor/src/lunar_interceptor/interceptor/traffic_filter.py
import logging
from re import compile, match
from typing import Dict, Optional, List
from socket import gethostbyname, error as socket_error
from ipaddress import ip_address, IPv4Address, IPv4Network

# This regex check test: https://regex101.com/r/Zt9EyZ/1
_NOT_ONLY_NUMBERS = compile(r"^[\d.]+$")
_RE_ADDRESS_VALIDATOR = compile(
    r"^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9]){2,}$"
)

# Consts
_LIST_DELIMITER = ","
_ALLOW_LIST = "AllowList"
_BLOCK_LIST = "BlockList"
_ALLOWED_HEADER_VALUE = "true"
_ALLOWED_HEADER_KEY = "x-lunar-allow"
_PRIVATE_IP_RANGES: Dict[str, IPv4Network] = {
    "10": IPv4Network("10.0.0.0/8"),
    "12": IPv4Network("127.0.0.0/8"),
    "17": IPv4Network("172.16.0.0/12"),
    "19": IPv4Network("192.168.0.0/16"),
}
_BLACK_HOLE = IPv4Network("0.0.0.0/32")


class TrafficFilter:
    """
    TrafficFilter Mechanism assists to filter the traffic forwarded through the Proxy.

    Args:
        raw_block_list (Optional[str]): The block list allow the user to avoid forwarding
                                                selected hosts or IPs through Lunar Proxy
        raw_allow_list (Optional[str]): The allow list allow the user to only forward
                                                selected hosts or IPs through Lunar Proxy
        logger (logging.Logger): Logger.
    """

    def __init__(
        self,
        raw_block_list: Optional[str],
        raw_allow_list: Optional[str],
        logger: logging.Logger,
    ):
        self._logger = logger
        self._managed = False
        self._is_external_cache: Dict[str, bool] = dict()
        self._block_list: Optional[List[str]] = self._parse_list(raw_block_list)
        self._allow_list: Optional[List[str]] = self._parse_list(raw_allow_list)
        self._state_ok = self.is_access_list_valid()
        self._logger.debug(
            f"TrafficFilter loaded, TrafficFilter validation passed successfully:\
                 {self._state_ok}"
        )

    def is_allowed(self, host_or_ip: str, headers: Optional[Dict[str, str]]) -> bool:
        """Check if the given HOST or IP should be forward through the Proxy

        Args:
            host_or_ip (str): The HOST or IP to check for.

        Returns:
            bool: True if the given HOST or IP should be forward through the Proxy.
        """
        if not self._state_ok:
            return False

        header_based_filter = self._check_for_header_based_filter(headers)
        if header_based_filter is not None:
            return header_based_filter

        return self._check_if_host_or_ip_is_allowed(host_or_ip=host_or_ip)

    @property
    def managed(self) -> bool:
        """Check if the Proxy is managed

        Returns:
            bool: True if Proxy is managed else False.
        """
        return self._managed

    @managed.setter
    def managed(self, managed: bool = False) -> None:
        """Update the status of whether the proxy is managed and validate the list after the update.

        Returns:
            None
        """
        self._logger.debug(f"Proxy is running in managed={managed} mode")
        self._managed = managed

    def _check_if_host_or_ip_is_allowed(self, host_or_ip: str) -> bool:
        """Check if the given HOST or IP should be forward through the Proxy

        Args:
            host_or_ip (str): The HOST or IP to check for.

        Returns:
            bool: True if the given HOST or IP should be forward through the Proxy.
        """
        is_allowed = self._check_allowed(host_or_ip=host_or_ip)
        if is_allowed is not None:
            return is_allowed

        return self._check_blocked(host_or_ip) and self._is_external(host_or_ip)

    def _check_allowed(self, host_or_ip: str) -> Optional[bool]:
        """Check if the given HOST or IP should be forward through the Proxy

        Args:
            host_or_ip (str): The HOST or IP to check for.

        Returns:
            bool: True if the given HOST or IP should be forward through the Proxy.
        """
        if self._allow_list is None:
            return None

        return host_or_ip in self._allow_list

    def _check_blocked(self, host_or_ip: str) -> bool:
        """Check if the given HOST or IP should not be forward through the Proxy

        Args:
            host_or_ip (str): The HOST or IP to check for.

        Returns:
            bool: True if the given HOST or IP should not be forward through the Proxy.
        """
        if not self._block_list:
            return True

        if host_or_ip in self._block_list:
            return False

        return True

    def _parse_list(self, value_to_parse: Optional[str]) -> Optional[List[str]]:
        """Parsed the given value_to_parse if needed.

        Args:
            value_to_parse (Optional[str]): A string to parse to a List of strings.

        Returns:
            Optional[List[str]]: If exists, returns a list of strings or None if empty.
        """
        if value_to_parse is None or not value_to_parse:
            return None

        return value_to_parse.split(_LIST_DELIMITER)

    def _check_for_header_based_filter(
        self, headers: Optional[Dict[str, str]]
    ) -> Optional[bool]:
        if not headers:
            return None

        is_allowed = headers.pop(_ALLOWED_HEADER_KEY, None)

        if is_allowed is None:
            return is_allowed

        return is_allowed == _ALLOWED_HEADER_VALUE

    def is_access_list_valid(self) -> bool:
        """Validates the destinations lists.

        Returns:
            bool: True if the validation flow passed successfully, False otherwise.
        """
        if not self._validate_allow():
            return False

        if not self._validate_block():
            self._logger.warning(
                "Interceptor will be disable to avoid \
                 passing wrong traffic through the Proxy."
            )
            return False

        return True

    def _validate_allow(self) -> bool:
        """Validate the allow list values

        Returns:
            bool: True if the validation passed,
                  in the allow list we dont failed on wrong values
                  and continue the flow.
        """
        if not self._allow_list:
            return True

        values_to_remove: List[str] = []

        for host in self._allow_list:
            if not (self._validate_host(host) or self._validate_ip(host)):
                values_to_remove.append(host)

        for value_to_remove in values_to_remove:
            self._logger.warning(
                f"Unsupported value '{value_to_remove}' will removed from the allowed list."
            )
            self._allow_list.remove(value_to_remove)

        return True

    def _validate_block(self):
        """Validate the block list values

        Returns:
            bool: True if the validation passed, False otherwise.
        """
        if not self._block_list:
            return True

        if self._allow_list:
            self._logger.warning(
                f"TrafficFilter::Found {_ALLOW_LIST} skipping the {_BLOCK_LIST}"
            )
            self._block_list = []
            return True

        block_list_validation_pass = True
        for host in self._block_list:
            if not (self._validate_host(host) or self._validate_ip(host)):
                self._logger.warning(
                    f"Error while parsing '{host}' from the block list"
                )
                block_list_validation_pass = False

        return block_list_validation_pass

    def _validate_host(self, host: str) -> bool:
        """Validate that the given host is valid.

        Args:
            host (str): The host to validate

        Returns:
            bool: True if the host is valid, False otherwise.
        """
        if match(_NOT_ONLY_NUMBERS, host) is not None:
            return False

        return match(_RE_ADDRESS_VALIDATOR, host) is not None

    def _is_external(self, host_or_ip: str) -> bool:
        """Check whether an HOST or IP is external or not

        Args:
            host_or_ip (str): The HOST or IP to check for.

        Returns:
            bool: True if the HOST or IP is external, False otherwise
        """
        is_external = self._is_external_cache.get(host_or_ip, None)

        if is_external is not None:
            return is_external

        if self._validate_ip(host_or_ip):
            is_external = self._is_external_ip(host_or_ip)

        else:
            is_external = self._is_external_domain(host_or_ip)

        if is_external is None:
            # For cases that we could not resolve the destination.
            # We do not store the result to try again on next iteration.
            return False

        self._is_external_cache[host_or_ip] = is_external
        return is_external

    def _validate_ip(self, ip: str) -> bool:
        """Validate that the given IP is a valid address

        Args:
            ip (str): The ip to validate

        Returns:
            bool: True if the IP is valid, False otherwise.
        """
        try:
            ip_address(ip)
            return True

        except:
            return False

    def _is_external_ip(self, ip: str) -> bool:
        """Check whether an IP is external or not

        Args:
            ip (str): The IP to check for.

        Returns:
            bool: True if the IP is external, False otherwise
        """
        return IPv4Address(ip) not in _PRIVATE_IP_RANGES.get(ip[:2], _BLACK_HOLE)

    def _is_external_domain(self, host: str) -> Optional[bool]:
        """Check whether an HOST is external or not

        Args:
            host (str): The HOST to check for.

        Returns:
            bool: True if the HOST is external, False otherwise
        """
        try:
            return self._is_external_ip(gethostbyname(host))

        except socket_error as error:
            # If there is a network error, we will avoid storing this and will try again next time.
            self._logger.warning(
                f"TrafficFilter::Could not resolve: '{host}'. Error: {error}"
            )
            return None
